unittests/osr_test.py

```diff
@@ -0,0 +1,153 @@
+              print('mousemove:'+e.pageX+","+e.pageY)
+            }, false);
+
+            document.addEventListener("dragenter", function( e ) {
+              print('dragenter:'+e.pageX+","+e.pageY)
+            }, false);
+
+            document.addEventListener("dragover", function( e ) {
+              print('dragover:'+e.pageX+","+e.pageY)
+            }, false);
+            document.addEventListener("drop", function( e ) {
+              print('drop:'+e.pageX+","+e.pageY)
+            }, false);
+
+
+        }
+        print('window.onload init')
+    </script>
+    <style>
+        .box{
+            background-color: #F00;
+            height: 200px;
+        }
+    </style>
+</head>
+<body>
+
+<input id="selectFile" type="file">
+<div class="box"></div>
+</body>
+</html>
+
+"""
+
+g_datauri = "data:text/html;base64," + base64.b64encode(g_datauri_data.encode(
+    "utf-8", "replace")).decode("utf-8", "replace")
+VIEWPORT_SIZE = (1024, 768)
+
+g_subtests_ran = 0
+
+def subtest_message(message):
+    global g_subtests_ran
+    g_subtests_ran += 1
+    print(str(g_subtests_ran) + ". " + message)
+    sys.stdout.flush()
+
+class Handler(object):
+    def __init__(self):
+        self.OnPaint_called = False
+
+    def OnConsoleMessage(self, message, **_):
+        if "error" in message.lower() or "uncaught" in message.lower():
+            self.javascript_errors_False = True
+            raise Exception(message)
+        else:
+            # Check whether messages from javascript are coming
+            self.OnConsoleMessage_True = True
+            subtest_message(message)
+
+    def OnDragEnter(self, browser, dragData, mask):
+        subtest_message('cef.OnDragEnter() OK')
+        return False
+
+    def StartDragging(self,browser, drag_data, allowed_ops, x, y):
+        subtest_message('cef.StartDragging() OK')
+        return False
+
+    def OnLoadEnd(self,browser,frame,http_code):
+        subtest_message('cef.OnLoadEnd() OK')
+        #testing mouse event
+        browser.SendMouseClickEvent(300, 20, cef.MOUSEBUTTON_LEFT, False, 1, cef.EVENTFLAG_LEFT_MOUSE_BUTTON)
+        browser.SendMouseMoveEvent(305, 25, False, cef.EVENTFLAG_LEFT_MOUSE_BUTTON)
+        browser.SendMouseClickEvent(305, 25, cef.MOUSEBUTTON_LEFT, True, 1, cef.EVENTFLAG_LEFT_MOUSE_BUTTON)
+
+        # testing drag event
+        # It can trigging cef.OnDragEnter, but still not trigger dragenter event in JS.
+        dragData = cef.DragData()
+        dragData.SetFragmentText('Test')
+        dragData.ResetFileContents()
+        browser.DragTargetDragEnter(dragData, 301, 21, cef.DRAG_OPERATION_COPY )
+        browser.DragTargetDragOver(302, 22, cef.DRAG_OPERATION_COPY )
+        browser.DragTargetDrop(303, 23)
+        browser.DragSourceEndedAt(303, 23, cef.DRAG_OPERATION_COPY )
+        browser.DragSourceSystemDragEnded()
+
+
+
+
+class OSRTest_IsolatedTest(unittest.TestCase):
+    def test_dragdrop(self):
```

Comment:
See this comment in main_test.py:
```
 def test_main(self):
        """Main entry point."""
        # All this code must run inside one single test, otherwise strange
        # things happen.
```

When working on main_test.py I noticed that you can't initialize CEF in one test run some other tests and then shutdown CEF in last test. I couldn't make it work with python's unittest module, strange things happened. The solution that worked was to run all tests inside on single test (test_main method) and introduce "subtest_message" function to report about results of various tests.

You can however run to separate tests by using IsolatedTest feature. See isolated_test.py - there are two classes defined IsolatedTest1 and IsolatedTest2 which run using separate python processes. This is the only way for CEF to work, because CEF can be initialized/shutdown only once during app lifetime.

I think the easiest thing is to define test_osr() and put both test_dragdrop() and test_dragdata() inside there. I don't mean to put all the code from test_dragdata() inside test_osr(), just make a call from there and the code should be in some separate function as it is now, but just don't make it a separate test. 

The test_dragdata code seems to work now, however this seems to be invalid. Most of CEF API is intended to be called only after CEF was initialized. In your case CEF is initialized in test_dragdrop(), however this is a separate test and it shuts down CEF at the end, so after shut down you are not allowed to call any of CEF API anymore. You shouldn't instantiate cef.DragData() aftert CEF was shut down.